fe/fe-core/src/main/java/com/starrocks/consistency/PartitionVersionRecovery.java

```diff
@@ -0,0 +1,174 @@
+        super("partition_version_recovery");
+    }
+
+    @Override
+    protected void runAfterCatalogReady() {
+        recover();
+    }
+
+    public void recover() {
+        GlobalStateMgr stateMgr = GlobalStateMgr.getCurrentState();
+        if (!stateMgr.getNodeMgr().getClusterInfo().checkTabletReportCacheUp(stateMgr.getDominationStartTimeMs())) {
+            LOG.warn("check tablet report time failed, wait for next round");
+            return;
+        }
+
+        List<PartitionVersion> partitionsToRecover = new ArrayList<>();
+        List<Long> dbIds = stateMgr.getLocalMetastore().getDbIds();
+        for (long dbId : dbIds) {
+            Database database = stateMgr.getDb(dbId);
+            if (database == null) {
+                continue;
+            }
+
+            if (database.isSystemDatabase()) {
+                continue;
+            }
+
+            Locker locker = new Locker();
+            locker.lockDatabase(database, LockType.READ);
+            try {
+                for (Table table : database.getTables()) {
+                    if (!table.isOlapTableOrMaterializedView()) {
+                        continue;
+                    }
+
+                    OlapTable olapTable = (OlapTable) table;
+
+                    for (Partition partition : olapTable.getAllPartitions()) {
+                        short replicaNum = olapTable.getPartitionInfo().getReplicationNum(partition.getId());
+
+                        PARTITION:
+                        for (PhysicalPartition physicalPartition : partition.getSubPartitions()) {
+                            long recoverVersion = -1L;
+                            for (MaterializedIndex idx : physicalPartition.getMaterializedIndices(
+                                    MaterializedIndex.IndexExtState.VISIBLE)) {
+                                for (Tablet tablet : idx.getTablets()) {
+                                    LocalTablet localTablet = (LocalTablet) tablet;
+                                    long quorumVersion = localTablet.getQuorumVersion(replicaNum / 2 + 1);
+                                    if (quorumVersion == -1L) {
+                                        LOG.warn("cannot find quorum version for tablet: {}, " +
+                                                "ignore partition: {}-{} table: {}-{} db: {}-{}",
+                                                tablet.getId(), physicalPartition.getId(), partition.getName(),
+                                                table.getId(), table.getName(), database.getId(), database.getFullName());
+                                        continue PARTITION;
+                                    }
+                                    if (recoverVersion == -1L) {
+                                        recoverVersion = quorumVersion;
+                                    } else if (recoverVersion != quorumVersion) {
+                                        LOG.warn("found different quorum versions: {} {}, " +
```

Comment:
normally, this shouldn't happen, right?

Reply:
Shadow index is an intermediate state for schema change. The tablets of shadow index maybe unhealthy at certain stages of schema change.